arrow-udf-python/src/lib.rs

```diff
@@ -293,6 +397,216 @@ impl Runtime {
+                row.clear();
+                row.push(state.clone());
+                for (column, field) in input.columns().iter().zip(input.schema().fields()) {
+                    let pyobj = self.converter.get_pyobject(py, field, column, i)?;
+                    row.push(pyobj);
+                }
+                let args = PyTuple::new_bound(py, row.drain(..));
+                let func = if ops.is_valid(i) && ops.value(i) {
+                    retract
+                } else {
+                    &aggregate.accumulate
+                };
+                state = func.call1(py, args)?;
+            }
+            let output = self
+                .converter
+                .build_array(&aggregate.state_field, py, &[state])?;
+            Ok(output)
+        })?;
+        Ok(new_state)
+    }
+
+    /// Merge states of an aggregate function.
+    ///
+    /// # Example
+    /// ```
+    #[doc = include_str!("doc_create_aggregate.txt")]
+    /// let states = Int32Array::from(vec![Some(1), None, Some(3), Some(5)]);
+    ///
+    /// let state = runtime.merge("sum", &states).unwrap();
+    /// assert_eq!(&*state, &Int32Array::from(vec![9]));
+    /// ```
+    pub fn merge(&self, name: &str, states: &dyn Array) -> Result<ArrayRef> {
+        let aggregate = self.aggregates.get(name).context("function not found")?;
+        let merge = aggregate.merge.as_ref().context("merge not found")?;
+        let output = self.interpreter.with_gil(|py| {
+            let mut state = self
+                .converter
+                .get_pyobject(py, &aggregate.state_field, states, 0)?;
+            for i in 1..states.len() {
+                if aggregate.mode == CallMode::ReturnNullOnNullInput && states.is_null(i) {
+                    continue;
+                }
+                let state2 = self
+                    .converter
+                    .get_pyobject(py, &aggregate.state_field, states, i)?;
+                let args = PyTuple::new_bound(py, [state, state2]);
+                state = merge.call1(py, args)?;
+            }
+            let output = self
+                .converter
+                .build_array(&aggregate.state_field, py, &[state])?;
+            Ok(output)
+        })?;
+        Ok(output)
+    }
+
+    /// Get the result of an aggregate function.
+    ///
+    /// If the `finish` function is not defined, the state is returned as the result.
+    ///
+    /// # Example
+    /// ```
+    #[doc = include_str!("doc_create_aggregate.txt")]
+    /// let states: ArrayRef = Arc::new(Int32Array::from(vec![Some(1), None, Some(3), Some(5)]));
+    ///
+    /// let outputs = runtime.finish("sum", &states).unwrap();
+    /// assert_eq!(&outputs, &states);
+    /// ```
+    pub fn finish(&self, name: &str, states: &ArrayRef) -> Result<ArrayRef> {
+        let aggregate = self.aggregates.get(name).context("function not found")?;
+        let Some(finish) = &aggregate.finish else {
+            return Ok(states.clone());
```

Comment:
What will happen if `state_type` is different from `output_type`?

Reply:
Oh, I saw this.

    /// - `finish(state) -> value`: Get the result of the aggregate function.
    ///     If not defined, the state is returned as the result. `output_type` must equal to `state_type`.

Personally I tend to make `finish()` required to avoid this type-unsafety 😄 Anyway, if users do want it, it's pretty easy to write a `finish()` to just return the state.

Reply:
I should add a check to ensure types are equal if `finish` is not defined. 😄

> Personally I tend to make `finish()` required to avoid this type-unsafety 😄 Anyway, if users do want it, it's pretty easy to write a `finish()` to just return the state.

This is how I designed at the beginning. But I realized that if the `finish()` is required, we have to call the python code, even the function is identity. If not required, we have the opportunity to return the array directly. The cost would be much less.